episodes/03-muscle-connect.md

```diff
@@ -0,0 +1,418 @@
+
+- the total simulation time to run this sub-model `tmax`
+- the time step to use `dt`
+- and the model parameter `k`
+
+Change the code such that we request these settings from the MUSCLE manager.
+
+:::::: solution
+
+```python
+from libmuscle import Instance
+from ymmsl import Operator
+
+def reaction(initial_state: np.array) -> np.array:
+    """
+    ...
+    """
+    instance = Instance({
+            Operator.F_INIT: ['initial_state'],       # np.array
+            Operator.O_F: ['final_state']})           # np.array
+
+    while instance.reuse_instance():
+        # begin F_INIT
+        t_max = instance.get_setting('t_max', 'float')
+        dt = instance.get_setting('dt', 'float')
+        k = instance.get_setting('k', 'float')
+
+        U = initial_state
+
+        t_cur = 0
+        # end F_INIT
+
+        # begin state_update_loop
+        ...
+```
+
+Note that getting settings needs to happen within the reuse loop; doing it before can lead to incorrect results.
+
+::::::
+:::
+
+## Receiving messages
+Apart from settings, we can use the `libmuscle.Instance.receive` function to receive an initial state for this submodel on the `initial_state` port. Note that we have declared that port above, and declared it to be associated with the `F_INIT` operator. During `F_INIT`, messages can only be received, not sent, so that declaration makes `initial_state` a receiving port.
+
+The message that we will receive can contain several pieces of information. For now, we are interested in the `data` and `timestamp` attributes. We assume the data to be a grid of floats containing our initial state and the time stamp tells us the simulated time at which this state is valid. We can store receive a message and store the `data` and `timestamp` attributes in the following way:
+
+```python
+    msg = instance.receive('initial_state')
+
+    data = msg.data.array.copy()
+    timestamp = msg.timestamp
+```
+
+::: callout
+The `msg.data` attribute holds an object of type `libmuscle.Grid`, which holds a read-only NumPy array and optionally a list of index names. Here, we take the array and make a copy of it for `data`, so that we can modify `data` in our upcoming state update. Without calling `.copy()`, the variable `data` would end up pointing to the same read-only array, and we would get an error message if we tried to modify it.
+
+The `timestamp` attribute is a double precision float containing the number of simulated (not wall-clock) seconds since the whole simulation started. Since `t_cur` is assigned the value of the timestamp, we don't need to make a copy.
+:::
+
+## State update loop
+
+The actual state update happens in the operator $S$ in the Submodel Execution Loop, and in our example model, it is determined entirely by the current state. Since no information from outside is needed, we do not receive any messages, and in our `libmuscle.Instance` declaration above, we did not declare any ports associated with `Operator.S`. 
```

Comment:
```suggestion
The actual state update happens in the operator $S$ in the Submodel Execution Loop and, in our example model, it is determined entirely by the current state. Since no information from outside is needed, we do not receive any messages, and in our `libmuscle.Instance` declaration above, we did not declare any ports associated with `Operator.S`. 
```